apiserver/facades/agent/keyupdater/authorisedkeys.go

```diff
@@ -52,100 +36,176 @@ var _ KeyUpdater = (*KeyUpdaterAPI)(nil)
+// the authorised keys for a requested tag.
+// - [params.CodeMachineInvalidID] When one of the machine tags id is invalid.
+// - [params.CodeMachineNotFound] When one of the machine's does not exist.
+func (api *KeyUpdaterAPI) AuthorisedKeys(
+	ctx context.Context,
+	arg params.Entities,
+) (params.StringsResults, error) {
 	if len(arg.Entities) == 0 {
 		return params.StringsResults{}, nil
 	}
 	results := make([]params.StringsResult, len(arg.Entities))
 
-	// For now, authorised keys are global, common to all machines.
-	var keys []string
-	config, configErr := api.model.ModelConfig(ctx)
-	if configErr == nil {
-		keys = ssh.SplitAuthorisedKeys(config.AuthorizedKeys())
-	}
-
-	controllerConfig, err := api.controllerConfigService.ControllerConfig(ctx)
-	if err == nil {
-		keys = append(keys, ssh.SplitAuthorisedKeys(controllerConfig.SystemSSHKeys())...)
-	} else if err != nil {
-		configErr = stderrors.Join(configErr, err)
-	}
-
 	canRead, err := api.getCanRead()
 	if err != nil {
-		return params.StringsResults{}, err
+		return params.StringsResults{}, fmt.Errorf(
+			"checking can read status for key updater authorised keys: %w",
+			err,
+		)
 	}
+
 	for i, entity := range arg.Entities {
 		tag, err := names.ParseTag(entity.Tag)
 		if err != nil {
-			results[i].Error = apiservererrors.ServerError(err)
+			results[i].Error = apiservererrors.ParamsError(
+				params.CodeTagInvalid,
+				"cannot parse tag %q",
+				entity.Tag,
+			)
 			continue
 		}
-		// 1. Check permissions
-		if !canRead(tag) {
-			results[i].Error = apiservererrors.ServerError(apiservererrors.ErrPerm)
+
+		if tag.Kind() != names.MachineTagKind {
+			results[i].Error = apiservererrors.ParamsError(
+				params.CodeTagKindNotSupported,
+				"tag %q unsupported, can only accept tags of kind %q",
+				tag, names.MachineTagKind,
+			)
 			continue
 		}
-		// 2. Check entity exists
-		if _, err := api.state.FindEntity(tag); err != nil {
-			if errors.Is(err, errors.NotFound) {
-				results[i].Error = apiservererrors.ServerError(apiservererrors.ErrPerm)
-			} else {
-				results[i].Error = apiservererrors.ServerError(err)
-			}
+
+		if !canRead(tag) {
+			results[i].Error = apiservererrors.ParamsError(
+				params.CodeUnauthorized,
+				"does not have permission to read authorised keys for %q",
+				tag,
+			)
 			continue
 		}
-		// 3. Get keys
-		if configErr == nil {
-			results[i].Result = keys
-		} else {
-			err = configErr
+
+		machineName := coremachine.Name(tag.Id())
+		keys, err := api.keyUpdaterService.GetAuthorisedKeysForMachine(ctx, machineName)
+
+		switch {
+		case errors.Is(err, errors.NotValid):
+			results[i].Error = apiservererrors.ParamsError(
+				params.CodeMachineInvalidID,
+				"invalid machine name %q",
+				machineName,
+			)
+			continue
+		case errors.Is(err, machineerrors.NotFound):
+			results[i].Error = apiservererrors.ParamsError(
+				params.CodeMachineNotFound,
+				"machine %q does not exist",
+				machineName,
+			)
+			continue
```

Comment:
As more and more api facades start to expect machine not found errors from service calls, we will end up with a lot of cut and paste boiler plate like this. It will very likely lead to skew in the error message depending on who writes the code; somewhere else, someone might use `machine %q not found` for example. Then our stuff looks untidy and inconsistent. Multiply out by all the different bespoke errors we will end up having and it gets worse. This code is really just splatting the internals of the helper we used to use all over the code base. I wonder if there's something we can improve here.

Reply:
Most definitely.

The reason I am breaking out of the helper methods:
- Helper methods have no context about the call site and in which order things should be checked.
- Helper methods do not have the context that the call site has. Generic handlers add more ambiguity for the reader as apposed to different formats.

Helper methods do give us consistency. But that consistency shouldn't be held in high regard here I think. The error messages you are generating at this level are still not user facing error messages and they shouldn't be treated as such. At best they error messages for log files in the off chance something slips through. The code should be transformed by the client into something meaningful and consistent.

This will also change per client as each client has a different master they trying to serve. So at this level I am still not concerned about the contents. I am more just trying to make sure that the context is there so that at the very worst case of understanding this the technical reader has as much context as possible.

</brain-dump>

Reply:
One thing that we can do thinking about this more is to look at an interface similar to `slog`. We already having the workings for this now.

We could have generic handlers for domain errors and then we could just be adding context attributes to the error. See https://pkg.go.dev/log/slog#Error for what I am thinking.

Reply:
Adding context from the caller to a generic params error constructor is fine and is what the current `apiservererrors.ServerError()` is missing I think. The other consideration is that the source `err.Error()` message would still be useful to have for logging/debugging. It's utility does become less though if we add context attributes in addition to the error code.

Reply:
I think  for the moment this discussion doesn't belong in this PR. I am going to leave it for the spec that will follow soon.